sycl/include/CL/sycl/detail/type_list.hpp

```diff
@@ -0,0 +1,141 @@
+//==----------- type_list.hpp - SYCL list of types utils -------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/access/access.hpp>
+#include <CL/sycl/detail/stl_type_traits.hpp>
+
+#include <type_traits>
+
+namespace cl {
+namespace sycl {
+namespace detail {
+
+template <typename T> using head_t = typename T::head;
+
+template <typename T> using tail_t = typename T::tail;
+
+// type_list
+template <typename... T> class type_list;
+
+using empty_type_list = type_list<>;
+
+template <typename T>
+struct is_empty_type_list
+    : conditional_t<std::is_same<T, empty_type_list>::value, std::true_type,
+                    std::false_type> {};
+
+template <> struct type_list<> {};
+
+template <typename H, typename... T> struct type_list<H, T...> {
+  using head = H;
+  using tail = type_list<T...>;
+};
+
+template <typename H, typename... T, typename... T2>
+struct type_list<type_list<H, T...>, T2...> {
+private:
+  using remainder = tail_t<type_list<H>>;
+  static constexpr bool has_remainder = !is_empty_type_list<remainder>::value;
+  using without_remainder = type_list<T..., T2...>;
+  using with_remainder = type_list<remainder, T..., T2...>;
+
+public:
+  using head = head_t<type_list<H>>;
+  using tail = conditional_t<has_remainder, with_remainder, without_remainder>;
+};
+
+// is_contained
+template <typename T, typename L>
+struct is_contained
+    : conditional_t<std::is_same<remove_cv_t<T>, head_t<L>>::value,
+                    std::true_type, is_contained<T, tail_t<L>>> {};
+
+template <typename T>
+struct is_contained<T, empty_type_list> : std::false_type {};
+
+// non_type_list
+template <typename T, T... V> struct non_type_list;
+
+template <typename T, T H, T... V> struct non_type_list<T, H, V...> {
+  static constexpr T head = H;
+  using tail = non_type_list<T, V...>;
+};
+
+template <typename T> struct non_type_list<T> {};
```

Comment:
How about naming this value_list instead?

Reply:
Done.